ranger/config/commands.py

```diff
@@ -1176,24 +1177,86 @@ def execute(self):
 
 
 class chmod(Command):
-    """:chmod <octal number>
+    """:chmod [<octal number>|<mode-spec>]
 
-    Sets the permissions of the selection to the octal number.
+    Sets the permissions of the selection to the given mode.
 
     The octal number is between 0 and 777. The digits specify the
     permissions for the user, the group and others.
 
     A 1 permits execution, a 2 permits writing, a 4 permits reading.
     Add those numbers to combine them. So a 7 permits everything.
+
+    The <mode-spec> is a comma-separated where each entry has the
+    form [ugoa]+[+-][rwx]+
     """
 
-    def execute(self):
-        mode_str = self.rest(1)
+    def __init__(self, *args, **kwargs):
+        '''init chmod-command.'''
+        super(chmod, self).__init__(*args, **kwargs)
+        self.re_mode = None
+        self.bitmasks = None
+
+    def prepare(self):
+        '''create regex for mode-spec and bitmasks.'''
+        self.re_mode = re.compile(r'''
+            ([ugoa]+)     # (u)ser, (g)roup, (o)other, (a)ll
+            ([+-])        # add (+), remove (-)
+            ([rwx]+)      # (r)ead, (w)rite, e(x)ecute
+            $             # end
+            ''', re.X)
+        masks = {
+            'u': {'r': stat.S_IRUSR, 'w': stat.S_IWUSR, 'x': stat.S_IXUSR},
+            'g': {'r': stat.S_IRGRP, 'w': stat.S_IWGRP, 'x': stat.S_IXGRP},
+            'o': {'r': stat.S_IROTH, 'w': stat.S_IWOTH, 'x': stat.S_IXOTH}}
+        masks['a'] = {
+            'r': masks['u']['r'] | masks['g']['r'] | masks['o']['r'],
+            'w': masks['u']['w'] | masks['g']['w'] | masks['o']['w'],
+            'x': masks['u']['x'] | masks['g']['x'] | masks['o']['x']}
+        self.bitmasks = masks
+
+    def get_mask_for(self, ugoa_part, rwx_part):
+        '''return mask for "ugoa" and "rwx" string.'''
+        mask = 0
+        for part in ugoa_part:
+            for flag in rwx_part:
+                mask |= self.bitmasks[part][flag]
+        return mask
+
+    def exec_modespecs(self, mode_str):
+        '''chmod according to mode-specs.'''
+        mask_operations = []
+        for mode in mode_str.split(','):
+            match = re.match(self.re_mode, mode)  # fullmatch missing in Py2
+            if match is None:
+                self.fm.notify(
+                    "Syntax: chmod <mode spec>: cannot parse " + mode,
+                    bad=True)
+                return False
+            sub_mask = self.get_mask_for(match.group(1), match.group(3))
+            mask_operations.append(
+                (lambda found, sub=sub_mask: found | sub)
```

Comment:
We don't actually use the second argument so let's just close over sub_mask.

Reply:
Also, I'd rather coalesce this into a single bitmask to apply.

Reply:
"single bitmask to apply" ... is not possible! 
Take `u-wx,u+r`  then I have for each bit *three*(!) possible actions: don't change, set or clear. There is *no* way to express this with a single bitmask that is applied. I express this with one `and` operation and an `or` operation (using lambda-Functions). 

Here is the [link to the gnu coreutils](https://fossies.org/linux/coreutils/lib/modechange.c#l_21): they are doing a similar approach because of this.

Reply:
Oh, good motivation then. I think we can still close over `sub_mask` though?